app/src/test/java/com/github/se/eduverse/repository/ProfileRepositoryTest.kt

```diff
@@ -602,8 +622,222 @@ class ProfileRepositoryImplTest {
+    whenever(mockTransaction.get(mockTargetDocRef)).thenReturn(mockTargetSnapshot)
+    whenever(mockTransaction.get(mockFollowerDocRef)).thenReturn(mockFollowerSnapshot)
+    whenever(mockTargetSnapshot.getLong("followers")).thenReturn(6L)
+    whenever(mockFollowerSnapshot.getLong("following")).thenReturn(4L)
+
+    whenever(mockFirestore.runTransaction<Void>(any())).thenAnswer { invocation ->
+      val transactionFunction = invocation.arguments[0] as Transaction.Function<Void>
+      transactionFunction.apply(mockTransaction)
+      Tasks.forResult(null)
+    }
+
+    // Execute toggleFollow
+    val result = repository.toggleFollow(followerId, targetId)
+
+    // Verify the result is false (no longer following)
+    assertFalse(result)
+
+    // Verify the unfollow operation
+    verify(mockDocRef).delete()
+
+    // Verify follow counts were updated correctly
+    verify(mockTransaction).update(mockTargetDocRef, "followers", 5L)
+    verify(mockTransaction).update(mockFollowerDocRef, "following", 3L)
+  }
+
+  @Test
+  fun `toggleFollow handles errors gracefully`() = runTest {
+    val followerId = "user1"
+    val targetId = "user2"
+    val mockQuery = mock(Query::class.java)
+
+    // Mock isFollowing to throw an exception
+    whenever(mockCollectionRef.whereEqualTo("followerId", followerId)).thenReturn(mockQuery)
+    whenever(mockQuery.whereEqualTo("followedId", targetId)).thenReturn(mockQuery)
+    whenever(mockQuery.get()).thenThrow(RuntimeException("Firestore error"))
+
+    try {
+      repository.toggleFollow(followerId, targetId)
+      fail("Expected an exception to be thrown")
+    } catch (e: Exception) {
+      // Verify that the exception is propagated
+      assertTrue(e is RuntimeException)
+      assertEquals("Firestore error", e.message)
+    }
+  }
+
+  @Test
+  fun `getProfile handles errors gracefully`() = runTest {
+    val userId = "testUser"
+
+    // Mock profile document to throw exception
+    val profileDocRef = mock(DocumentReference::class.java)
+    whenever(mockCollectionRef.document(userId)).thenReturn(profileDocRef)
+    whenever(profileDocRef.get()).thenThrow(RuntimeException("Firestore error"))
+
+    try {
+      repository.getProfile(userId)
+      fail("Expected an exception to be thrown")
+    } catch (e: Exception) {
+      assertTrue(e is RuntimeException)
+      assertEquals("Firestore error", e.message)
+    }
+
+    verify(profileDocRef).get()
+  }
+
   @After
   fun tearDown() {
     Dispatchers.resetMain() // Reset main dispatcher after the test
+    unmockkAll()
+  }
+}
+
+class TestableProfileRepositoryImpl(firestore: FirebaseFirestore, storage: FirebaseStorage) :
+    ProfileRepositoryImpl(firestore, storage) {
+  // Make isFollowing public so we can override it in tests
+  public override suspend fun isFollowing(followerId: String, targetUserId: String): Boolean {
+    return super.isFollowing(followerId, targetUserId)
```

Comment:
The ProfileRepositoryImplTest class is well structured and covers use cases like tracing, transaction management, and error handling effectively. Using mockk to simulate Firestore is smart and allows testing complex interactions. Some possible improvements include simplifying the repeated setup of mocks and adding comments for clarity. Overall, test coverage and robustness are solid.